crates/miden-objects/src/block/partial_account_tree.rs

```diff
@@ -0,0 +1,329 @@
+        PartialAccountTree { smt: PartialSmt::new() }
+    }
+
+    /// Returns a new [`PartialAccountTree`] instantiated with the provided entries.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if:
+    /// - the merkle paths of the witnesses do not result in the same tree root.
+    /// - there are multiple witnesses for the same ID _prefix_.
+    pub fn with_witnesses(
+        witnesses: impl IntoIterator<Item = AccountWitness>,
+    ) -> Result<Self, AccountTreeError> {
+        let mut tree = Self::new();
+
+        for witness in witnesses {
+            tree.track_account_witness(witness)?;
+        }
+
+        Ok(tree)
+    }
+
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns an opening of the leaf associated with the `account_id`. This is a proof of the
+    /// current state commitment of the given account ID.
+    ///
+    /// Conceptually, an opening is a Merkle path to the leaf, as well as the leaf itself.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if:
+    /// - the account ID is not tracked by this account tree.
+    pub fn open(&self, account_id: AccountId) -> Result<AccountWitness, AccountTreeError> {
+        let key = AccountTree::id_to_smt_key(account_id);
+
+        self.smt
+            .open(&key)
+            .map(|proof| {
+                // Check which account ID this proof actually contains. We rely on the fact that the
+                // tree only contains zero or one entry per account ID prefix.
+                //
+                // If the requested account ID matches an existing ID's prefix but their suffixes do
+                // not match, then this witness is for the _existing ID_.
+                //
+                // Otherwise, if the ID matches the one in the leaf or if it's empty, the witness is
+                // for the requested ID.
+                let witness_id = match proof.leaf() {
+                    SmtLeaf::Empty(_) => account_id,
+                    SmtLeaf::Single((key_in_leaf, _)) => {
+                        // SAFETY: By construction, the tree only contains valid IDs.
+                        AccountTree::smt_key_to_id(*key_in_leaf)
+                    },
+                    SmtLeaf::Multiple(_) => {
+                        unreachable!(
+                            "account tree should only contain zero or one entry per ID prefix"
+                        )
+                    },
+                };
+
+                let commitment = Digest::from(
+                    proof
+                        .get(&AccountTree::id_to_smt_key(witness_id))
+                        .expect("we should have received a proof for the witness key"),
+                );
+
+                // SAFETY: The proof is guaranteed to have depth AccountTree::DEPTH.
+                AccountWitness::new_unchecked(witness_id, commitment, proof.into_parts().0)
```

Comment:
Should this logic go into `AccountWitness`? It feels like this could be something like `AccountWitness::from_smt_proof()`.